partiql-parser/src/parse/partiql.lalrpop

```diff
@@ -15,7 +15,44 @@ use partiql_source_map::location::{ByteOffset, BytePosition, ToLocated};
 grammar<'input, 'err>(input: &'input str,
                       errors: &'err mut Vec<ErrorRecovery<ByteOffset, lexer::Token<'input>, ParseError<'input, BytePosition>>>);
 
+
+// ------------------------------------------------------------------------------ //
+//                                                                                //
+//                            UNION/INTERSECT/EXCEPT                              //
+//                                                                                //
+// ------------------------------------------------------------------------------ //
 pub Query: Box<ast::Expr> = {
+    <Query02>
+}
+
+Query02: Box<ast::Expr> = {
+    <lo:@L> <lhs:Query02> "UNION"  <setq:SetQuantifier> <rhs:Query01> <hi:@R> => {
+        Box::new( ast::Expr{ kind: ast::ExprKind::Union(ast::Union{setq, lhs, rhs}.ast(lo..hi)) } )
+	},
+    <lo:@L> <lhs:Query02> "EXCEPT"  <setq:SetQuantifier> <rhs:Query01>  <hi:@R> => {
+        Box::new( ast::Expr{ kind: ast::ExprKind::Except(ast::Except{setq, lhs, rhs}.ast(lo..hi)) } )
+	},
+    <Query01>
+}
+
+Query01: Box<ast::Expr> = {
+    <lo:@L> <lhs:Query01> "INTERSECT" <setq:SetQuantifier> <rhs:SingleQuery>  <hi:@R> => {
+        Box::new( ast::Expr{ kind: ast::ExprKind::Intersect(ast::Intersect{setq, lhs, rhs}.ast(lo..hi)) } )
+	},
+    <SingleQuery>
+}
+
+SetQuantifier: ast::SetQuantifier = {
+    "DISTINCT" => ast::SetQuantifier::Distinct,
+    "ALL"? => ast::SetQuantifier::All,
```

Comment:
I'm still trying to understand the syntax/semantics for lalrpop. What does the `?` denote? And should it be the same for both `DISTINCT` and `ALL`?

Reply:
LALRPOP refers to it as a macro.

 Four macros are built-in and have regex-like semantics: `?` is optional, `+` is one or more, `*` is zero or more, `(`...`)` is grouping. These correspond to returning the rust types `?` => `Option<T>`, `+`, `*` => `Vec<T>`

http://lalrpop.github.io/lalrpop/tutorial/006_macros.html